allennlp/common/sqlite_sparse_sequence.py

```diff
@@ -0,0 +1,102 @@
+import os
+import shutil
+from os import PathLike
+from typing import MutableSequence, Any, Union, Iterable
+
+from sqlitedict import SqliteDict
+
+
+class SqliteSparseSequence(MutableSequence[Any]):
+    def __init__(self, filename: Union[str, PathLike], read_only: bool = False):
+        self.table = SqliteDict(filename, "sparse_sequence", flag="r" if read_only else "c")
+
+    def __del__(self):
+        self.close()
+
+    def __getitem__(self, i: Union[int, slice]) -> Any:
+        if isinstance(i, int):
+            try:
+                return self.table[str(i)]
+            except KeyError:
+                current_length = len(self)
+                if i >= current_length or current_length <= 0:
+                    raise IndexError("list index out of range")
+                elif i < 0 < current_length:
+                    return self.__getitem__(i % current_length)
+                else:
+                    return None
+        elif isinstance(i, slice):
+            from allennlp.tango.dataloader import ShuffledSequence
+
+            return ShuffledSequence(self, range(*i.indices(len(self))))
+        else:
+            raise TypeError(f"list indices must be integers or slices, not {i.__class__.__name__}")
+
+    def __setitem__(self, i: Union[int, slice], value: Any):
+        if isinstance(i, int):
+            current_length = len(self)
+            if i < 0:
+                i %= current_length
+            self.table[str(i)] = value
+            self.table["_len"] = max(i, current_length)
+            self.table.commit()
+        else:
+            raise TypeError(f"list indices must be integers, not {i.__class__.__name__}")
```

Comment:
Why can't we handle slices here?

Reply:
You can't set slices. You can in numpy, but not in Python `list`.

Reply:
But can't you just go through one-by-one?

Reply:
I guess you can? I tried to do whatever `list` does. Numpy can do many things that `list` can't, and I didn't implement those either.

Reply:
Ok, I don't see a huge need for it, so 🤷‍♂️